### ISSUE TITLE
Om-internal-sound and memory

### ISSUE THREAD

**Issue by charlesneimog** [cac-t-u-s/om-sharp-users#32]
Hi again Jean,

I have a question.

Is it possible to clear the memory that om-sharp is using after save the <om-internal-sound> or <sound> in the hard-drive?

I am working in a piece, and I am organize samples with voices, but how my voice is now I little big, I am very often receive the message 

**Error while evaluationg the box sound-seq-list: failed to allocate foreign object of size 6584208 [allocation static]** 

When I evaluate the patch more that once.

The object sound-seq-list work with this code: 

```
(defun sound-seq-list (sounds fade-bewteen-sound)

(let*  (
    (action1 (sound-seq (first sounds) (second sounds) fade-bewteen-sound))
    (action2 (if 
                (>  (length (x-append action1 sounds)) 2)
                (x-append action1 (last-n sounds (- (length sounds) 2)))
                action1)))
    
    (if (< (length action2) 2)
            (first action2)
            (setf sounds (build-sound-sequence-fun action2 fade-bewteen-sound)))))
```
Maybe this code it's not good. I dont know.

Could you help me?

Thank you!

**Comment by j-bresson** [comment]
Hallo Charles.
In principle the audio buffer memory of the sounds created with `sound-seq` is automatically cleared when the sounds are garbage-collected.
What does `build-sound-sequence-fun` do ?

**Comment by charlesneimog** [comment]
It works like sound-seq but allows me to work with more than 2 sounds. 

For example in a list `(#<sound 2369E843> #<sound 2369B61F> #<sound 2369E653> #<sound 2369F5A7> #<sound 236A04C3> #<sound 2369B9A3>)`, `build-sound-sequence-fun` will build one sound #<sound 238745A7> that contains all the sounds of the list `(#<sound 2369E843> #<sound 2369B61F> #<sound 2369E653> #<sound 2369F5A7> #<sound 236A04C3> #<sound 2369B9A3>)`. in sequence. 

But here is my problem, after I evaluate this function and save this new sound, the memory is not cleared. I think that is because I do not know how do this 'garbage-collected'. 

Maybe I must to indicate to OM# that this memory is not important after save the sound? But I do not know how I do that.

**Comment by j-bresson** [comment]
Can you try with `reduce` ?

This Lisp function allows you to apply a function to a list, applying each result to the successive next elements.
It just requires passing `sound-seq` as a lambda function:

<img width="189" alt="Screenshot 2021-05-29 at 23 16 47" src="https://user-images.githubusercontent.com/7666142/120084907-7694e000-c0d4-11eb-9a08-c59064b47be8.png">

Let me know if the same memory issues arise with it.

(ps. I was interested to see the  code of `build-sound-sequence-fun`)

**Comment by charlesneimog** [comment]
Yes, I already tried it, and I had the same problem. 
I already tried:

- Save the sound after the processing without lock the boxes. 
- Use the reduce with sound-seq (It is a little bit more slow in my computer).
- And now I am using the defmethod!

```
(defmethod! sound-seq-list ((sound list) &optional (fade-bewteen-sound 0.001))
:initvals ' (NIL)
:indoc ' ("Sdif-File.")
:icon 'om-ckn
:doc ""

(om-print "Aguarde alguns segundos!" "Verbose")

(let* (
      (action1 (build-sound-sequence-fun sound fade-bewteen-sound)))

  (save-sound action1 (outfile (string+ "temp-sound" (list->string-fun (list (om::om-random 0 9999999))) ".wav")))))
```


Sorry, the code is that above, I forgot to change the name. 

```
(defun build-sound-sequence-fun (sounds fade-bewteen-sound)

(let*  (
    (action1 (sound-seq (first sounds) (second sounds) fade-bewteen-sound))
    (action2 (if 
                (>  (length (x-append action1 sounds)) 2)
                (x-append action1 (last-n sounds (- (length sounds) 2)))
                action1)))
    
    (if (< (length action2) 2)
            (first action2)
            (setf sounds (build-sound-sequence-fun action2 fade-bewteen-sound)))))
```

**Comment by j-bresson** [comment]
Ok, I think you could simplify your recursion code a little bit, but this is not the point here.
Are you using big audio files ? Only 6 sounds of reasonable/small  sizes shouldn't be a problem...  
I could suggest a few things:
- Force-call garbage collection using `(gc-all)` : does it make a difference ?
- Compile your function: `(compile #'build-sound-sequence-fun)`

**Comment by charlesneimog** [comment]
Hi Jean, 

It does not work, `gc-all` clear just 10MB of ram. I changed my work to make the sequences and mix with SOX, based in the OM-SOX. It's working now.

Thank you!

**Comment by j-bresson** [comment]
Good that you got a solution with om-sox!

I'd like to keep investigating this: What is the approximate size of audio files you are using ?

OM# on Windows is a 32-bit application, so it has limited resources. Also, recursive processes tend to consume a lot of these since audio buffers remain allocated on the stack until the recursion returns. Loops are usually more efficient in this respect.

**Comment by charlesneimog** [comment]
I am working with small samples (like 900Kb) but put it in sequence (the sequence is the problem), so in the end, my audio has a size of 15Mb or more.

**Comment by charlesneimog** [comment]
One more question/curiosity, why is it not possible to compile OM# in 64bits?

**Comment by j-bresson** [comment]
> I am working with small samples (like 900Kb) but put it in sequence (the sequence is the problem), so in the end, my audio has a size of 15Mb or more.

On the example above there seemed to be only 6 samples in the sequence, right ? 🤔

**Comment by j-bresson** [comment]
> One more question/curiosity, why is it not possible to compile OM# in 64bits?

💰  &rarr; http://www.lispworks.com/buy/prices-2a.html

**Comment by charlesneimog** [comment]
> > I am working with small samples (like 900Kb) but put it in sequence (the sequence is the problem), so in the end, my audio has a size of 15Mb or more.
> 
> On the example above there seemed to be only 6 samples in the sequence, right ? 🤔

Yes, but in my real work I am working with a lot of samples. In my actual work I have 146 samples. And is with this work that I am having some problems. 

Normally with 2 evaluations I already have the message with allocation problem.

![work](https://user-images.githubusercontent.com/31707161/121830204-e2fa0c80-cc9a-11eb-923b-6ec67a1d4a9b.PNG)
